### PR TITLE
#3902: validate unknown properties in @Ignored

### LINKED ISSUE #3902: Unknown fields in `@Ignored` lead to an internal error
Add an unknown field like '"foo"' to the [ZooMapper](https://github.com/mapstruct/mapstruct/blob/6b6600c370cb50b2fbf40f56412238974575ae85/processor/src/test/java/org/mapstruct/ap/test/ignored/ZooMapper.java#L21) results in an error:

```
CompilationResultDescriptor [compilationResult=FAILED, diagnostics=[DiagnosticDescriptor: ERROR ZooMapper.java:13 Internal error in the mapping processor: java.lang.NullPointerException: Cannot invoke "javax.lang.model.element.Element.equals(Object)" because "e" is null
at org.mapstruct.ap.internal.processor.MapperAnnotatedFormattingMessenger.methodInMapperClass(MapperAnnotatedFormattingMessenger.java:140)
at org.mapstruct.ap.internal.processor.MapperAnnotatedFormattingMessenger.determineDelegationElement(MapperAnnotatedFormattingMessenger.java:136)
at org.mapstruct.ap.internal.processor.MapperAnnotatedFormattingMessenger.printMessage(MapperAnnotatedFormattingMessenger.java:69)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.handleDefinedMapping(BeanMappingMethod.java:1310)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.handleDefinedMappings(BeanMappingMethod.java:1175)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.build(BeanMappingMethod.java:296)
at org.mapstruct.ap.internal.model.AbstractBaseBuilder.lambda$createForgedAssignment$1(AbstractBaseBuilder.java:92)
at org.mapstruct.ap.internal.model.AbstractBaseBuilder.getOrCreateForgedAssignment(AbstractBaseBuilder.java:108)
at org.mapstruct.ap.internal.model.AbstractBaseBuilder.createForgedAssignment(AbstractBaseBuilder.java:95)
at org.mapstruct.ap.internal.model.PropertyMapping$PropertyMappingBuilder.forgeMapping(PropertyMapping.java:832)
at org.mapstruct.ap.internal.model.PropertyMapping$PropertyMappingBuilder.forgeMapping(PropertyMapping.java:796)
at org.mapstruct.ap.internal.model.PropertyMapping$PropertyMappingBuilder.forge(PropertyMapping.java:323)
at org.mapstruct.ap.internal.model.PropertyMapping$PropertyMappingBuilder.build(PropertyMapping.java:270)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.applyPropertyNameBasedMapping(BeanMappingMethod.java:1624)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.applyPropertyNameBasedMapping(BeanMappingMethod.java:1586)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.build(BeanMappingMethod.java:314)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.getMappingMethods(MapperCreationProcessor.java:432)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.getMapper(MapperCreationProcessor.java:189)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.process(MapperCreationProcessor.java:143)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.process(MapperCreationProcessor.java:86)
at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:373)
```

This should lead to a correct error.

### PR DESCRIPTION
**Pull request by znight1020**
Fixes #3902

**Solution**
The root cause of the `NullPointerException` was identified within the `BeanMappingMethod.handleDefinedMapping` method's error reporting logic.

**Root Cause Analysis**
1. For each property string in `@Ignored(targets = {"foo"})`, MapStruct internally creates a MappingOptions object. When processing an unknown property like "foo", the call to `mapping.getElement()` returns `null`.
2. The processor then attempts to report the "Unknown property" error by calling `Messager.printMessage()`, passing the null element. This results in an NPE, as the messager requires a valid source element to determine the error location.

**Implementation Details**
1. The fix addresses this by modifying the error handling logic for unknown properties within handleDefinedMapping:
A new, more specific error message `BEANMAPPING_UNKNOWN_PROPERTY_IN_IGNORED` was added to `Message.java`.
2. Before reporting an error, the logic now uses a specific condition `(mapping.isIgnored() && mapping.getElement() == null)` to detect properties that originate from an `@Ignored` annotation and do not exist. This correctly distinguishes the case from a standard `@Mapping(target="...", ignore=true)` on an unknown property.
3. When the mapping element is null, the error is reported on the method itself (using its ExecutableElement). This prevents the original NPE and provides a clear diagnostic to the user at the method level.
  - ex) No property named "foo" exists in @Ignored for target type "Issue3902Mapper.ZooDto"

With these changes, all mvn clean install builds now pass successfully.

### CODE REVIEW

File: integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationExclusionCliEnhancer.java [hunk 1]
```diff
@@ -27,6 +27,7 @@ public Collection<String> getAdditionalCommandLineArguments(ProcessorTest.Proces
         additionalExcludes.add( "org/mapstruct/ap/test/bugs/_1596/*.java" );
         additionalExcludes.add( "org/mapstruct/ap/test/bugs/_1801/*.java" );
         additionalExcludes.add( "org/mapstruct/ap/test/bugs/_3089/*.java" );
+        additionalExcludes.add( "org/mapstruct/ap/test/bugs/_3902/*.java" );
```
Comment (thunderhook) [inline integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationExclusionCliEnhancer.java]: I don't think that this is needed. I uncommented that line and my maven build was successful. Could you please take another look at it?

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 2]
```diff
@@ -45,6 +45,7 @@ public enum Message {
     BEANMAPPING_CYCLE_BETWEEN_PROPERTIES( "Cycle(s) between properties given via dependsOn(): %s." ),
     BEANMAPPING_UNKNOWN_PROPERTY_IN_DEPENDS_ON( "\"%s\" is no property of the method return type." ),
     BEANMAPPING_IGNORE_BY_DEFAULT_WITH_MAPPING_TARGET_THIS( "Using @BeanMapping( ignoreByDefault = true ) with @Mapping( target = \".\", ... ) is not allowed. You'll need to explicitly ignore the target properties that should be ignored instead." ),
+    BEANMAPPING_UNKNOWN_PROPERTY_IN_IGNORED("Unknown property \"%s\" in @Ignored for target type \"%s\". No property named \"%s\" exists in the source or target."),
```
Comment (thunderhook) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Since `@Ignored` is only about ignoring properties of the **target** we should not mention "source" here.
What about something like this?

```suggestion
    BEANMAPPING_UNKNOWN_PROPERTY_IN_IGNORED("No property named \"%s\" exists in @Ignored for target type \"%s\""),
```

**Review by thunderhook — CHANGES_REQUESTED** [review]
Hi @znight1020!

Thanks a lot for your PR!  I have two minor comments, but other than that, great work!

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 2 above)
Comment (znight1020) [reply processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Thank you for the suggestion!
You are right, `@Ignored` is indeed target-specific, so the new message is much more accurate.

I will apply the change and update the messages in the test case accordingly.

File: integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationExclusionCliEnhancer.java (hunk 1 above)
Comment (znight1020) [reply integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationExclusionCliEnhancer.java]: When I run a `mvn clean install` locally, I'm seeing a `CompilationFailureException` during `MavenIntegrationTest.fullFeatureTest()` with the following error:

```
[ERROR] .../mapstruct/processor/src/test/java/org/mapstruct/ap/test/bugs/_3902/Issue3902Mapper.java:[23,12] No property named "foo" exists in @Ignored for target type "Issue3902Mapper.ZooDto"
```

I suspect this is because I'm running the build which includes the `integrationtest` module.

I noticed the comment in `MavenIntegrationTest` that mentions automatic exclusion for files named with `*Erroneous*`.

```
/**
 * ... excluding all classes that contain ... in their path/file name:
 * ...
 * <li>{@code *Erroneous*}</li>
 * ...
 */
```
Would the correct approach here be to rename `Issue3902Mapper.java` to `ErroneousIssue3902Mapper.java` and remove the explicit exclusion I added?

Comment (thunderhook) [reply integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationExclusionCliEnhancer.java]: Good catch!
Yes that would indeed be better, the `FullFeatureCompilationExclusionCliEnhancer` should be used to exclude tests for specific compilers (and it seems for some edge cases using SPIs or accessor naming strategies). 

(and my comment "I uncommented that line and my maven build was successful" was a lie, I did not run the integration test module, sorry for that 😅)

Comment (znight1020) [reply integrationtest/src/test/java/org/mapstruct/itest/tests/FullFeatureCompilationExclusionCliEnhancer.java]: That's totally fine! Your comment was the perfect hint that led me to discover the Erroneous naming convention, so it was very helpful 😄

I've applied the requested changes. Thank you again for the detailed review and great guidance!

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java [hunk 3]
```diff
@@ -1280,40 +1280,54 @@ else if ( inheritContext.isReversed() ) {
...
+                    if ( elementForMessage == null ) {
+                        elementForMessage = method.getExecutable();
+                    }
+
+                    if ( mapping.isIgnored() && mapping.getElement() == null ) {
+                        msg = Message.BEANMAPPING_UNKNOWN_PROPERTY_IN_IGNORED;
                         args = new String[] {
                             targetPropertyName,
-                            resultTypeToMap.describe(),
-                            mostSimilarProperty
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:1294]: I don't understand why we removed the `mostSimilarProperty` from this error message?

**Review by filiphr — COMMENTED** [review]
Thanks @znight1020 for the fix.

What was the reason for not using " Did you mean \"%s\"?"  in the `BEANMAPPING_UNKNOWN_PROPERTY_IN_IGNORED` error message?

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java (hunk 3 above)
Comment (znight1020) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:1294]: To be honest, my main focus was on fixing the NullPointerException and ensuring a clear, specific error message for `@Ignored` was reported. I completely overlooked adding the "Did you mean" suggestion, my apologies for that.

Adding it would be much more consistent with the other 'Unknown property' errors. Should I create a new PR for this enhancement?

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:1294]: It would be great if you could create another PR and add that as well
